internal/nginx/modules/httpmatches_test.js

```diff
@@ -0,0 +1,243 @@
+    });
+    it('returns 404 if not all args exist in request', function () {
+      r.args = {
+        Arg1: 'value1',
+        arg2: 'value2=SOME=other=value',
+      };
+      redirect.redirect(r);
+      expect(r.testReturned).to.equal(404);
+    });
+    it('returns 404 if all args exist in request but not all match', function () {
+      // r has all headers but not all values match
+      r.args = {
+        Arg1: 'value1',
+        arg2: 'value2=SOME=other=value',
+        ARg3: '==value3&*1(*+', // query param matching is case sensitive, so this shouldn't match.
+      };
+      redirect.redirect(r);
+      expect(r.testReturned).to.equal(404);
+    });
+    it('redirects to path if all args exist in request and all match', function () {
+      // r has all headers but not all values match
+      r.args = {
+        Arg1: 'value1',
+        arg2: 'value2=SOME=other=value',
+        arg3: '==value3&*1(*+',
+      };
+      redirect.redirect(r);
+      expect(r.testRedirectedTo).to.equal('/params');
+    });
+  });
+  describe('multi-condition matching', function () {
+    beforeEach(function () {
+      let matches = {
+        method: 'GET',
+        headers: ['header1:value1', 'header2:value2'],
+        params: ['Arg1=value1', 'arg2=value2=SOME=other=value'],
+        path: '/a-match',
+      };
+      r.variables = {
+        http_matches: JSON.stringify(matches),
+      };
+    });
+    it('returns 405 if method does not match', function () {
+      // method doesn't match but everything else does
+      r.method = 'POST';
+      r.headersIn = {
+        header1: 'value1',
+        header2: 'value2',
+      };
+      r.args = {
+        Arg1: 'value1',
+        arg2: 'value2=SOME=other=value',
+      };
+      redirect.redirect(r);
+      expect(r.testReturned).to.equal(405);
+    });
+    it('returns 404 headers do not match', function () {
+      // method and args match, but headers don't
+      r.method = 'GET';
+      r.headersIn = {
+        header1: 'value1',
+      };
+      r.args = {
+        Arg1: 'value1',
+        arg2: 'value2=SOME=other=value',
+      };
+      redirect.redirect(r);
+      expect(r.testReturned).to.equal(404);
+    });
+    it('returns 404 if args do not match', function () {
+      // method and headers match, but args don't
+      r.method = 'GET';
+      r.headersIn = {
+        header1: 'value1',
+        header2: 'value2',
+      };
+      r.args = {
+        arg2: 'value2=SOME=other=value',
+      };
+      redirect.redirect(r);
+      expect(r.testReturned).to.equal(404);
+    });
+    it('redirects to path if all conditions match', function () {
+      r.method = 'GET';
+      r.headersIn = {
+        header1: 'value1',
+        header2: 'value2',
+      };
+      r.args = {
+        Arg1: 'value1',
+        arg2: 'value2=SOME=other=value',
+      };
+      redirect.redirect(r);
+      expect(r.testRedirectedTo).to.equal('/a-match');
+    });
+  });
+});
```

Comment:
would it make sense to test for cases when the http_match is malformed for headers and params?

Reply:
Added a few tests for malformed headers and params